jgrapht-core/src/main/java/org/jgrapht/alg/cycle/AhujaOrlinSharmaCyclicExchangeLocalAugmentation.java

```diff
@@ -0,0 +1,450 @@
+ * as published by the Free Software Foundation, or (at your option) any
+ * later version.
+ *
+ * or (per the licensee's choosing)
+ *
+ * (b) the terms of the Eclipse Public License v1.0 as published by
+ * the Eclipse Foundation.
+ */
+package org.jgrapht.alg.cycle;
+
+import org.jgrapht.Graph;
+import org.jgrapht.graph.GraphWalk;
+import org.jgrapht.util.TypeUtil;
+
+import java.util.*;
+
+/**
+ * Implementation of an algorithm for the local augmentation problem for the cyclic exchange neighborhood,
+ * i.e. it finds subset-disjoint negative cycles in a graph, based on
+ * Ravindra K. Ahuja, James B. Orlin, Dushyant Sharma,
+ * A composite very large-scale neighborhood structure for the capacitated minimum spanning tree problem,
+ * Operations Research Letters, Volume 31, Issue 3, 2003, Pages 185-194, ISSN 0167-6377,
+ * https://doi.org/10.1016/S0167-6377(02)00236-5. (http://www.sciencedirect.com/science/article/pii/S0167637702002365)
+ *
+ * A subset-disjoint cycle is a cycle such that no two vertices in the cycle are in the same subset of a given partition of the whole vertex set.
+ *
+ * This algorithm returns the first found negative subset-disjoint cycle, that is, the cycle has minimum number of vertices. It may enumerate all paths up
+ * to the length given by the parameter <code>lengthBound</code>, i.e the algorithm runs in exponential time.
+ *
+ * This algorithm is used to detect valid cyclic exchanges in a cyclic exchange neighborhood for the Capacitated Minomum Spanning Tree problem {@link org.jgrapht.alg.spanning.AhujaOrlinSharmaCapacitatedMinimumSpanningTree} @see org.jgrapht.alg.spanning.AhujaOrlinSharmaCapacitatedMinimumSpanningTree
+ *
+ * @param <V> the vertex type
+ * @param <E> the edge type
+ *
+ * @author Christoph Grüne
+ * @since June 7, 2018
+ */
+public class AhujaOrlinSharmaCyclicExchangeLocalAugmentation<V, E> {
+
+    /**
+     * the input graph
+     */
+    private Graph<V, E> graph;
+    /**
+     * the map that maps each vertex to a subset (identified by labels) of the partition
+     */
+    private Map<V, Integer> labelMap;
+    /**
+     * bound on how long the cycle can get
+     */
+    private int lengthBound;
+    /**
+     * contains whether the best or the first improvement is returned
+     */
+    private boolean bestImprovement;
+
+    /**
+     * Constructs an algorithm with given inputs
+     *
+     * @param graph the (improvement) graph on which to calculate the local augmentation
+     * @param lengthBound the (inclusive) upper bound for the length of cycles to detect
+     * @param labelMap the labelMap of the vertices encoding the subsets of vertices
+     * @param bestImprovement contains whether the best or the first improvement is returned: best if true, first if false
+     */
+    public AhujaOrlinSharmaCyclicExchangeLocalAugmentation(Graph<V, E> graph, int lengthBound, Map<V, Integer> labelMap, boolean bestImprovement) {
```

Comment:
I am still missing the structure of the `graph` parameter here. I would add a section in the Javadoc where I would describe the exact structure of the improvement graph. 

Otherwise the class should be considered an implementation helper class which means that it should be private.

Reply:
I improved the constructor. I am not sure, whether I understand you correctly; you probably have a problem with `(improvement)` in the description of the graph parameter. Actually, the graph has to be directed (at least in this implementation; it is very probable that it can be easily extended to undirected graphs as well) and the vertices have to be partitioned. Then, the definition of the subset-disjoint cycle actually says what the output is. There no other restrictions and improvement graphs are only a sub class of the above mentioned type. That is why I deleted `(improvement)` in the description.